felix/fv/qos_controls_test.go

```diff
@@ -156,6 +164,106 @@ var _ = infrastructure.DatastoreDescribe(
 
 			})
 		})
+
+		Context("With packet rate limits", func() {
+			getRules := func(felixId int) string {
+				out, err := tc.Felixes[felixId].ExecOutput("iptables-save", "-c")
+				log.Infof("iptables-save -c output:\n%v", out)
+				Expect(err).NotTo(HaveOccurred())
+				return out
+			}
+
+			It("should limit packet rate correctly", func() {
+				format.MaxLength = 1000000
+
+				By("Starting iperf3 server on workload 0")
+				serverCmd := w[0].ExecCommand("iperf3", "-s")
+				err := serverCmd.Start()
+				Expect(err).NotTo(HaveOccurred())
+
+				By("Running iperf3 client on workload 1")
+				out, err := w[1].ExecOutput("iperf3", "-c", w[0].IP, "-O5", "-M1000", "-J")
+				Expect(err).NotTo(HaveOccurred())
+				baselineRate, err := getRateFromJsonOutput(out)
+				Expect(err).NotTo(HaveOccurred())
+				log.Infof("iperf client rate with no packet rate limit (bps): %v", baselineRate)
+				// Expect the baseline rate to be much greater (>=100x) the bandwidth that we
+				// would get with the packet rate we are going to configure just below. In
+				// practice we see several Gbps here.
+				Expect(baselineRate).To(BeNumerically(">=", 800000.0*100))
+
+				By("Setting 100 packets/s limit for ingress on workload 0 (iperf3 server)")
+				w[0].WorkloadEndpoint.Spec.QoSControls = &api.QoSControls{
+					IngressPacketRate: 100,
+				}
+				w[0].UpdateInInfra(infra)
+				Eventually(tc.Felixes[0].ExecOutputFn("ip", "r", "get", "10.65.0.2"), "10s").Should(ContainSubstring(w[0].InterfaceName))
+
+				By("Waiting for the config to appear in 'iptables-save' on workload 0")
+				// ingress config should be present
+				Eventually(getRules(0), "10s", "1s").Should(And(MatchRegexp(`-A cali-tw-`+regexp.QuoteMeta(w[0].InterfaceName)+` .* -m comment --comment "Mark packets within ingress packet rate limit" -m limit --limit `+regexp.QuoteMeta("100/sec")+` -j MARK --set-xmark 0x\d+/0x\d+`), MatchRegexp(`-A cali-tw-`+regexp.QuoteMeta(w[0].InterfaceName)+` .* -m comment --comment "Drop packets over ingress packet rate limit" -m mark ! --mark 0x\d+/0x\d+ -j DROP`)))
```

Comment:
In this check, `iptables-save -c` will be called only once, returning a string, and then that same string will be analysed up to 10 times.  If it fails the first time, it won't make any difference to keep trying for the next 10s.

What you really want is:
```
			getRules := func(felixId int) func() string {
                                return func() {
				out, err := tc.Felixes[felixId].ExecOutput("iptables-save", "-c")
				log.Infof("iptables-save -c output:\n%v", out)
				Expect(err).NotTo(HaveOccurred())
				return out
			}
			}
```